### DISCUSSION TITLE
Button remap

### DISCUSSION THREAD

**Discussion by n222384** [nk64/GT4286#9] [General]
I've followed the guide and edited the FF FF FF FF on the main file and replaced the keymap file however the Med Kick (B) and Hard Kick (L) buttons are non-responsive on player one. Player 2 works fine.

What am I doing wrong? I just want to be able to play Street Fighter 2 Hyper Fighting...  :-(

**Comment by Ryhawkian** [comment]
Yup i did the edits as well and no luck

**Reply by horsiezoo** [reply]
strange, it works fine here, I renamed the new map file to fbneo.keymap and replaced existing file, unsure if thats right/not but seems to work ok

**Reply by Ryhawkian** [reply]
the Keymap works fine. but the P1 Button 5/6 aren't working even with doing the hex edit

**Reply by horsiezoo** [reply]
![IMG_2064](https://github.com/user-attachments/assets/e36a8e9b-ccd6-4fab-9699-1eabf5d2d89c)
These ones? Eg: R L ?

**Reply by Ryhawkian** [reply]
with the original FBNeo those LR don't work, you can remap them with the keymap file so those work then the buttons you remaped from don't work.. ATM the default P1 only has 4 working buttons. Supposedly if you hex edit the fbneo you can fix this, but can't seem to get it working

**Reply by horsiezoo** [reply]
Can you pls try with a different SF eg: SFA 3 ? Also try replacing with https://github.com/nk64/GT4286/tree/main/sdcard_tweaks/emus/mame  -- unsure if that's patched or not but worth a try

**Reply by n222384** [reply]
No. The bottom two ( B and L) are unresponsive after the remap and hexedit.

Buttons do work in the fb neo main menu.

**Reply by nk64** [reply]
> Can you pls try with a different SF eg: SFA 3 ? Also try replacing with https://github.com/nk64/GT4286/tree/main/sdcard_tweaks/emus/mame -- unsure if that's patched or not but worth a try

FYI: The ```fbneo``` file you can find there is an example shell script which is a starting place to explore the inner workings of the linux operating system of the device. It is not a patched emulator.

**Comment by nk64** [comment]
@Ryhawkian @n222384
Thanks for giving it a go.

Please note that:
- the emulator file to patch is ```/emus/fbneo/fbneo```
- if you also intend to use the "MAME" emulator even though it is a duplicate you can also patch ```/emus/mame/fbneo```
- there are [6 bytes to patch spread across three locations](https://github.com/nk64/GT4286/blob/main/img/fbneo-hexedit-fix-p1-r-and-l-buttons.png)

Here's my thinking on how I would troubleshoot your problem:
1. restore the original emulator file(s) (from the backup you took)
2. confirm that you are using the emulators from a 157 game generation SD card. You can use the [G4268Util](https://github.com/nk64/GT4286/releases) to identify your card (if it is relatively stock).
3. check that the SHA 256 Hash of your emulator file matches ```24B4026764C6B8F218D56222A534EDB3D21210A2C8FCDEE8A5F96505D6B85E9F```
4. Edit ```/emus/fbneo/fbneo``` modify the 6 bytes (in three locations)
5. Check that the SHA 256 Hash of the emulator file now matches ```BB9DA0976667A3656EB851582B3D02D49253983E8A10FCBDBEAA47EB25F74282```
6. Repeat the procedure for the 'MAME' version in ```/emus/mame/fbneo```
7. Boot the console and load a 6 button game like ```Street Fighter II``` test that the P1-L and P1-R buttons now respond in a FBA game (and optionally in MAME). At this stage the buttons probably won't be in a nice order.
8. Copy (and rename) the ```fbneo.keymap.sf2+patchedemu``` file found [here](https://github.com/nk64/GT4286/tree/main/sdcard_tweaks/keyremap) to ``/keyremap/fbneo.keyremap``` (backed this one up first too)
9. Boot the console and try the arcade version of ```Street Fighter II``` and check that all six buttons are in the right order.
10. Report back your results (and let me know if I've missed any steps) :-)

**Reply by tim-mcg** [reply]
@nk64 Thank you - I can confirm that these instructions work. The only difference from what I was doing before was starting with a 157 game SD card instead of the 1619 game one. Note: the SHA-256 of the fbneo file was actually the same on both cards, so maybe there's a difference somewhere else that's preventing the fix from working.

**Reply by n222384** [reply]
Sorted it.

SHA256 Hash was checking out but turns out I was only editing the file under /emus/fbneo/fbneo

Need to also edit the file under /emus/mame/fbneo

Love the work and I'm having more fun playing around with the system than actually playing the games (although others are doing all the heavy lifting - I'm just following the instructions here)

**Reply by Ryhawkian** [reply]
+1 to the making sure the one under mame was edited. that fixed it and agreed with n222384 thanks for all the help

**Reply by nk64** [reply]
@tim-mcg Thanks Tim, on closer inspection I concur that the emulators are all the same across the currently know generations, only the menu has changed. I can't say how meaningfully it has changed, but one thing I have noticed about the older cards is that the games that have an single quote in their file name were stored in the database with that single quote replaced with a double quote character, but then when they are displayed in the game list the have a single quote. My utility fixes that (invisible curiosity) up and the games still seem to work fine.

There may yet be some situations where single quotes and dots might still cause issue but hopefully we can work around them and if not document the problems.

**Reply by tim-mcg** [reply]
Ok, so it must have just been user error. I have started with a clean 1619 card and carefully followed the steps and it's working perfectly. I still have my original files on my hard drive and both fbneo and mame were actually edited. It may just have been as simple as not copying the edited mame file onto the SD card.

**Comment by ot644** [comment]
Can confirm that this works perfectly. Need to ensure all bytes in the screenshot (both lines) are modified.

**Comment by freeb1e4me** [comment]
From my experience, remapping the keys work great for the add-on roms like Street Fighter, but it will mess up the buttons for multi-button games that are built into the 1500+ games. For example, breakers (a street fighter clone) which uses 4 buttons is messed up. Hopefully, someone else can confirm this.

If true, a possible solution is to use separate remap keys for roms/fbneo and roms/mame. Put add-on roms into roms/mame and use the remapped keys and keep original keys for rom/fbneo. No idea how to do this though

**Reply by nk64** [reply]
Because the emulator used for mame games (```/emus/mame/fbneo``` is a copy of the ```fbneo``` emulator, it uses the same keyremap file (```/keyremap/fbneo.keymap```).

We can patch the binary file in some way so that it looks for a different keymap file.

Just taking a quick look at emulator under Ghidra, it looks like there are two places we would need to patch.
I'll put it on the todo list.

**Comment by horsiezoo** [comment]
![399128278-e36a8e9b-ccd6-4fab-9699-1eabf5d2d89c](https://github.com/user-attachments/assets/016c5289-fba1-4160-81de-8a9bff9b62ba)
I found a few sega, snes and gb button test roms, Sega megadrive mapping is all over the place, as per photo , looking into how to remap that ATM

**Reply by horsiezoo** [reply]
![snesc](https://github.com/user-attachments/assets/cb55411d-98c1-45ce-a6ad-6d023d176201)
SNES map

**Reply by n222384** [reply]
Might be easier just to print out some cardboard templates with cutout holes to place over the buttons.

One cutout for each console.

**Reply by horsiezoo** [reply]
heh yes I used a whiteboard marker, but yes actually great idea!

**Reply by horsiezoo** [reply]
![psxc](https://github.com/user-attachments/assets/057a03df-eb7c-4bc2-9ece-cff36c1573a3)
PSX mappings default, I used PaRappa the Rapper to figure this out lol

**Reply by horsiezoo** [reply]
So along with almost everything else on this console someone obviously thought "how do we set this up in the most awkward annoying way possible ? We've been told to use all these surplus media player boards in an arcade machine and by the way has anyone in the office actually played a video game before"

**Comment by freeb1e4me** [comment]
*** another issue related to the 6 button patch ***

The reason why I messed with the controller settings is because the 6 button games keymap messes with the 4 button games, and not realizing it applies to all games (not game specific). Then there are games like Mortal Kombat which has a totally different 6 button arrangement (two buttons in the middle are 'block').

> I'm imagining a folder called `/roms/FBA/keymaps` with keymap files matching the original rom name eg `sf2.keymap` and a `1942.keymap` etc. Any games that have a keymap override get it copied over `/keyremap/fbneo.keymap` otherwise something like `_default.keymap` gets copied over then the emulator gets loaded.

1. Look for a specific game keymap in /roms/FBA/keymaps/specific_game.keymap. If the specific game keymap is found then use it, otherwise use the default in /keyremap/fbneo.keymap
2. When a game gets a keymap override in controller settings then copy that keymap into /roms/FBA/keymaps/specific_game.keymap - replacing the specific_game.keymap if one already exists. Then reset that override keymap back to the original from a backup in /keyremap/fbneo.keymap.bak

**Reply by n222384** [reply]
Maybe start off small and just do a keymap for each emulator. Replace the main emulator program with a script to copy the keymap for that emu (e.g. snes or MD) over the main keymap file. 

Will still have problems across the arcade games but at least the keys will then be consistent on each console/emulator

**Reply by horsiezoo** [reply]
Is there a .cfg /key map that actually works for Sega on this ? (Not for lack of trying, my /picodrive folder is full of differently named .cfg files in hope one of them is recognised, but no)

**Reply by nk64** [reply]
As far as I can see, Out of the box you can do a remap file per emulator.

Just load a game en each emulator and then open the emulator menu and  in to then muck around in the cobtroller settings page.
The result will (should) be stored in something like /keyremap/{emuname}.keymap (typing from memory, actual paths may vary)

the key remap is a small json file, but x in the json file doesn't necessarily the console's X button so there is a lot of methodical trial and error involved (I mostly hand edited the json file in my trial and error rather than using the controller configuration screen)

**Reply by horsiezoo** [reply]
> a lot of methodical trial and error involved

Sums up the week with this thing 

*ahem* you couldn't paste your mapping configs here could you by chance ? :-)

**Reply by horsiezoo** [reply]
Working Sega key map :
pico.keymap

{"a":"b","r1":"a","b":"x","x":"y","y":"l1","l1":"r1"}

-X--Y--R
-A--B--L
mapped to:
-x--y--z
-a--b--c

**Reply by freeb1e4me** [reply]
mortal kombat keymap:

{"x":"b","y":"a","a":"y","b":"b","l1":"x","l2":"r1","r1":"l1","r2":"r2"}

**Reply by horsiezoo** [reply]
Cool thx, so that just leaves PSX mapping, getting close to getting it right

**Reply by nk64** [reply]
I only got SF2 to work and then called it a day.

It looks like FBNeo has some logic specifically for CPS2 Street Fighter Style games to switch the layout. 
Check out the ```usesStreetFighterLayout()``` function here: https://github.com/finalburnneo/FBNeo/blob/master/src/burner/sdl/input_sdl.cpp

So for now (and for FBNeo) we can say that we need one mapping for the CPS2 Street Fighter games and another mapping for all the rest. 

If we figure out why the other consoles that FBNeo supports aren't working then the situation might change.

**Reply by horsiezoo** [reply]
PSX 

pcsx.keymap

{"a":"b","y":"y","x":"x","b":"a"}

many thanks nk64 your boot screen disable thing really helped here 

used https://www.psx-place.com/resources/psxtest.395/  to help with remap

-X--Y--R
-A--B--L
mapped to:
-[]--^--R1
-X--O--l1

**Reply by freeb1e4me** [reply]
I don't see a pcsx.keymap to replace.

In /keyremap, I can only see fbneo.keymap and fceux.keymap.

Also what's the difference the between the classes PCE and PS as I think they use the fbneo.keymap as well?

**Reply by horsiezoo** [reply]
Fbneo and pcsx Pico drive etc are completely separate emulators. If you boot a game in say PS (PlayStation) and press start and select and remap controls it’ll save a pcsx.keymap file ./keyremap that keymap is likely going to be incorrect. So the above should replace that

**Reply by freeb1e4me** [reply]
Thanks that makes sense. Do PS and PCE classes run from the same pcsx4all emulator?

**Reply by nk64** [reply]
> Thanks that makes sense. Do PS and PCE classes run from the same pcsx4all emulator?

No, they are different. see here for the various emulator classes: https://github.com/nk64/GT4286/blob/main/Emulators.md

**Reply by ot644** [reply]
> mortal kombat keymap:
> 
> {"x":"b","y":"a","a":"y","b":"b","l1":"x","l2":"r1","r1":"l1","r2":"r2"}

Awesome thanks... hopefully there will be a way to apply these keymaps to roms in a future release of G4268Util (no pressure nk64)..

**Reply by horsiezoo** [reply]
Basically if it helps think of this like a crap clone of retroarch without any of the useful stuff

**Comment by ot644** [comment]
I made some changes to the /emus/fbneo script to use the street fighter and mortal kombat key remap files provided. My keyremap directory looks like this.


fbneo.keymap.default (original keymap - not a big fan of the default layout)
fbneo.keymap.mk (mortal kombat keymap)
fbneo.keymap.sf2 (street fighter keymap)
fbneo.keymap


I added the following to the script before the wrapped executable is executed.


case "${redirectedfilename}" in
    *sf*)
        log "Detected 'sf' in redirected filename. Copying fbneo.keymap.sf2 to fbneo.keymap."
        cp /mnt/extsd/keyremap/fbneo.keymap.sf2 /mnt/extsd/keyremap/fbneo.keymap
        ;;
    *mk*)
        log "Detected 'mk' in redirected filename. Copying fbneo.keymap.mk to fbneo.keymap."
        cp /mnt/extsd/keyremap/fbneo.keymap.mk /mnt/extsd/keyremap/fbneo.keymap
        ;;
    *)
        log "No specific keymap detected. Copying fbneo.keymap.default to fbneo.keymap."
        cp /mnt/extsd/keyremap/fbneo.keymap.default /mnt/extsd/keyremap/fbneo.keymap
        ;;
esac


It works well for what I need.  Example logs below.

Mortal Kombat:
Redirected Rom: '/mnt/extsd/roms/fba/Mortal Kombat .redir' -> '/mnt/extsd/roms/fba/redir/mk.zip'
Detected 'mk' in redirected filename. Copying fbneo.keymap.mk to fbneo.keymap.
Executing wrapped exe: /mnt/extsd/emus/fbneo/fbneo.bin with arguments: /mnt/extsd/roms/fba/redir/mk.zip
Executing wrapped exe without logging: /mnt/extsd/emus/fbneo/fbneo.bin


Street Fighter:
Redirected Rom: '/mnt/extsd/roms/fba/Street Fighter II - The World Warrior .redir' -> '/mnt/extsd/roms/fba/redir/sf2.zip'
Detected 'sf' in redirected filename. Copying fbneo.keymap.sf2 to fbneo.keymap.
Executing wrapped exe: /mnt/extsd/emus/fbneo/fbneo.bin with arguments: /mnt/extsd/roms/fba/redir/sf2.zip
Executing wrapped exe without logging: /mnt/extsd/emus/fbneo/fbneo.bin


Other roms:
Redirected Rom: '/mnt/extsd/roms/fba/Altered Beast .redir' -> '/mnt/extsd/roms/fba/redir/altbeast.zip'
No specific keymap detected. Copying fbneo.keymap.default to fbneo.keymap.
Executing wrapped exe: /mnt/extsd/emus/fbneo/fbneo.bin with arguments: /mnt/extsd/roms/fba/redir/altbeast.zip
Executing wrapped exe without logging: /mnt/extsd/emus/fbneo/fbneo.bin


Maybe we can look at a new layout for the default keymap if anyone has any suggestions?

**Reply by horsiezoo** [reply]
Hey that’s pretty good! Was thinking along those lines, clever about matching text.
So annoying the .dats are hardcoded in flash, you know the one thing that actually could do with regular updates but no.
Being able to run:
https://archive.org/download/fbnarcade-fullnonmerged/arcade/test240p.zip
would help a lot.

On that is there enough Linux on this to be able redirect the .dat dir to sd card?

**Reply by ot644** [reply]
Below is the full modified script for anyone that is interested.  Obviously it can be modified for any particular rom/keymap as needed.  Don't forget to copy the required keymap files to the keyremap directory for the script to work.  A new layout for the default keymap would be ideal if someone can suggest.

[fbneo.txt](https://github.com/user-attachments/files/18323893/fbneo.txt)

**edit. attached script as txt**

**Reply by freeb1e4me** [reply]
Something strange is happening to the fbneo.keymap.default. When this keymap gets copied to fbneo.keymap the file gets corrupted. I tried multiple times and the same thing happens. This is what the keymap contents show:

≻≸∺≢∬≹∺≡∬≡∺≹∬≢∺≢∬ㅬ㨢砢Ⱒ氢∲∺ㅲⰢ爢∱∺ㅬⰢ爢∲∺㉲索਍

**Reply by horsiezoo** [reply]
Unix or windows formatted text file?

also run a chkdsk on your sd card

**Reply by ot644** [reply]
Does it only occur when fbneo.keymap.default is copied and not fbneo.keymap.sf2 or fbneo.keymap.mk? If so, try replacing the fbneo.keymap.default with one of the other working files and rename it to .default and test again.  If all good just replace the contents of the working file with the required default keymap.

**Reply by horsiezoo** [reply]
Did you figure out what was up with the corruption?

**Reply by freeb1e4me** [reply]
Yes, chkdsk did find an issue but it wasn't the problem as fixing it didnt help. Eventually, I worked it out to be a combination of a wrong file encoding (something like UTF 8 BOM) on the keymap.default and using notepad, which was not saving the file correctly probably due to the encoding. So, I switched to notepad++ and saved all the files to UNIX end of line endings. All fixed now, bye bye another 3 hrs of my life on this $50 unit.

**Reply by horsiezoo** [reply]
Hah yes I know what you mean, glad you worked it out. At least we now know how to fix that kind of issue
